### PR TITLE
BF: Removed allow_break

### PR DESCRIPTION
**Pull request by deka27**
# Removed allow_break from TissueClassifierHMRF

## Overview
This PR introduces removal of 'allow_break' from the TissueClassifierHMRF as mentioned in issue #3096. 

## Changes
- Removed the `allow_break` variable to simplify loop exit logic.
- The `tolerance` parameter now explicitly disables the early stopping condition when set to `0`, allowing for clear differentiation between using and not using the tolerance check for loop control.
- Clarified documentation for the `tolerance` parameter to indicate how users can disable the tolerance check.

## Related Issues
* #3096 

## Screenshots
I ran this [tutorial](https://docs.dipy.org/stable/examples_built/segmentation/tissue_classification.html#sphx-glr-examples-built-segmentation-tissue-classification-py) to check if the changes were functional. Here are the results.

Old Code

Screenshot 2024-03-18 at 7 12 32 PM

New Code

Screenshot 2024-03-18 at 7 09 19 PM

## Additional Comments
* I also ran the test mentioned in `test_mrf.py` by running the function `test_classify`, the test was successful.
```python
pytest -svv -k 'test_classify'
```
* Based on my understanding I have put the changes, but please let me know if this is the right way or any other changes are required. Any feedbacks would be helpful.

## Checklist
- [x] I have reviewed my own code changes.
- [x] I have tested my changes locally.
- [x] My code follows the project's coding standards.

### CODE REVIEW

File: dipy/segment/tissue.py [hunk 1]
```diff
@@ -5,57 +5,54 @@
...
-                prevent the ICM loop to stop. Default is 1e-05.
-        max_iter : float,
-                fixed number of desired iterations. Default is 100.
-                If the user only specifies this parameter, the tolerance
-                value will not be considered. If none of these two
-                parameters
+            Value that defines the percentage of change tolerated to
+            prevent the ICM loop to stop. Default is 1e-05.
+            If you want tolerance check to be disabled put 'tolerance = 0'.
+        max_iter : int,
```
Comment (skoudoro) [inline dipy/segment/tissue.py:40]: Can you replace  `int,` by `int, optional`

File: dipy/segment/tissue.py [hunk 2]
```diff
@@ -5,57 +5,54 @@
...
+            3D structural image.
         nclasses : int,
-                number of desired classes.
+            Number of desired classes.
         beta : float,
-                smoothing parameter, the higher this number the smoother the
-                output will be.
+            Smoothing parameter, the higher this number the smoother the
+            output will be.
         tolerance: float,
```
Comment (skoudoro) [inline dipy/segment/tissue.py:36]: `float, optional` instead, can you add it?

File: dipy/segment/tissue.py (hunk 1 above)
Comment (pjsjongsung) [inline dipy/segment/tissue.py:40]: Could you also edit the max_iter description?

**Review by skoudoro — COMMENTED** [review]
Hi @deka27,

Thank you for this, waiting for the CI's but this change make sense to me. 
see below few comments

**Comment by deka27** [comment]
I will be squash the commits now.

**Comment by deka27** [comment]
The squashing is done, Please let me know if things are correct now!

**Comment by deka27** [comment]
I have done the squashing as required, please let me know if this is correct.

**Review by skoudoro — APPROVED** [review]
LGTM. 

is there any additional comment? @pjsjongsung ? @arokem ?

**Comment by skoudoro** [comment]
Thanks @deka27 for this work.

**Comment by deka27** [comment]
> Thanks @deka27 for this work.

My Pleasure.

File: dipy/segment/tissue.py [hunk 3]
```diff
@@ -5,57 +5,58 @@
...
-        max_iter : float,
-                fixed number of desired iterations. Default is 100.
-                If the user only specifies this parameter, the tolerance
-                value will not be considered. If none of these two
-                parameters
+            Smoothing parameter, the higher this number the smoother the
+            output will be.
+        tolerance: float, optional
+            Value that defines the percentage of change tolerated to
+            prevent the ICM loop to stop. Default is 1e-05.
```
Comment (jhlegarreta) [inline dipy/segment/tissue.py:38]: Remove defaults here: this is difficult to maintain.

File: dipy/segment/tissue.py [hunk 4]
```diff
@@ -5,57 +5,58 @@
...
+            If you want tolerance check to be disabled put 'tolerance = 0'.
+        max_iter : int, optional
+            Fixed number of desired iterations. Default is 100.
```
Comment (jhlegarreta) [inline dipy/segment/tissue.py:41]: Remove defaults here: this is difficult to maintain.

File: doc/doc-requirements.txt [hunk 5]
```diff
@@ -15,7 +15,7 @@ numpydoc
 sphinx!=4.4
 sphinx-gallery>=0.10.0
 tomli>=2.0.1
-pydata-sphinx-theme==0.13.3
+grg-sphinx-theme>=0.3
```
Comment (jhlegarreta) [inline doc/doc-requirements.txt:18]: Why was this changed? If this was required, should have been done on a separate PR: this is unrelated to the `allow_break` issue. Also, this was part of another commit: https://github.com/dipy/dipy/commit/c2c42c44569ec12a4cdf5ed8a97c45bb114d6b11

File: dipy/segment/tissue.py [hunk 6]
```diff
@@ -81,51 +82,36 @@ def classify(self, image, nclasses, beta, tolerance=None, max_iter=None):
...
                 self.energies.append(energy)
-                if allow_break:
-                    self.energies_sum.append(energy_sum[-1])
-                else:
-                    self.energies_sum.append(energy[energy > -np.inf].sum())
-
-            if allow_break and i > 5:
+                self.energies_sum.append(energy_sum[-1])
 
+            if tolerance > 0 and i > 5:
```
Comment (jhlegarreta) [inline dipy/segment/tissue.py:110]: For a separate PR: if 5 means something here, it should be put to a variable and explain the meaning, as it is also used in https://github.com/dipy/dipy/pull/3132/files#diff-540e9ecc7f5c0befbaff5a9def01d422445705275835db8a00e8c763792e6e12R114.

File: dipy/segment/tissue.py [hunk 7]
```diff
@@ -5,57 +5,58 @@
...
+            explicitly set to 0, this early stopping mechanism is disabled,
+            and the algorithm will run for the specified number of
+            iterations unless another stopping criterion is met.
 
         Returns
         -------
         initial_segmentation : ndarray,
-                3D segmented image with all tissue types
-                specified in nclasses.
+            3D segmented image with all tissue types specified in nclasses.
```
Comment (jhlegarreta) [inline dipy/segment/tissue.py:53]: Style changes (and there are a lot here) should be on their own commit. Otherwise the review gets a lot harder, the real issue being addressed getting lost.

**Review by jhlegarreta — COMMENTED** [review]
@skoudoro @pjsjongsung Sorry for the late review: the scope of this was a little bit too broad IMO.

File: doc/doc-requirements.txt (hunk 5 above)
Comment (pjsjongsung) [reply doc/doc-requirements.txt:18]: Agree. It might have been from squashing some unrelated commit

File: dipy/segment/tissue.py (hunk 3 above)
Comment (pjsjongsung) [reply dipy/segment/tissue.py:38]: The default values were following what it already had. If we just remove the default values, it could result in breaking many user code in the next version. Plus, tolerance value is not something a user would change a lot, and is hard to understand how to select the appropriate value.

File: dipy/segment/tissue.py (hunk 6 above)
Comment (pjsjongsung) [reply dipy/segment/tissue.py:110]: Agree. We could give it a name like tolerance_window

File: dipy/segment/tissue.py (hunk 7 above)
Comment (pjsjongsung) [reply dipy/segment/tissue.py:53]: Agree. We need to double check if this change follows the styles of other code files as well.

File: dipy/segment/tissue.py (hunk 3 above)
Comment (jhlegarreta) [reply dipy/segment/tissue.py:38]: Only meant removing defaults in the documentation, same as in e.g.
https://github.com/dipy/dipy/pull/2707#discussion_r1071309787

If for some reason the default changes in the method signature, changes that the docstring goes out of sync are high.

Comment (pjsjongsung) [reply dipy/segment/tissue.py:38]: Makes sense, especially since it is not a value user would have to understand most of the time.

File: dipy/segment/tissue.py (hunk 6 above)
Comment (deka27) [reply dipy/segment/tissue.py:110]: I can put up a separate PR for that. Thank you for the feedback.

File: doc/doc-requirements.txt (hunk 5 above)
Comment (deka27) [reply doc/doc-requirements.txt:18]: Yea, sorry about that. I updated the branch from upstream while working on the issue and then squashed it, because of that this also went into the PR.
